### PR TITLE
Unify `@cypher` resolution for Union fields.

### LINKED ISSUE #4617: Authorization rules are not applied in nested Cypher fields
**Describe the bug**
In the issue https://github.com/neo4j/graphql/issues/326 there was described a bug where the authorization was not applied for custom Cypher fields. For the top-level Cypher field, the User authorization is applied correctly, but if you query the User from another entity then the authorization is no longer applied.

**Type definitions**
```graphql
type Post {
    title: String
    likedBy: [User!]!
        @cypher(
            statement: """
            MATCH (user:User)
            RETURN user
            """
            columnName: "user"
        )
}

type User {
    id: ID
    email: String!
        @authorization(validate: [{ when: [BEFORE], operations: [READ], where: { node: { id: "$jwt.sub" } } }])
}
```

**To Reproduce**
Execute the following query:

```graphql
{
    posts {
        likedBy {
            email
        }
    }
}
```

**Expected behavior**
I should receive a Forbidden error with an Invalid JWT and the correct results with a valid one.
Authorization rules should be applied consistently between top-level and nested.

**System (please complete the following information):**

* OS: [macOS]
* Version: [4.4.5]
* Node.js version: [20.11.0]

### PR DESCRIPTION
**Pull request by MacondoExpress**
Follow up on the PR: https://github.com/neo4j/graphql/pull/4734.
This PR contains the migration of the `@cypher` translation for Composite targets. This PR contains also the migration of the `CypherPropertySort`.

Closes: #4617

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3404 | 3404 | 37 | 48 | 1134 |
|  aggregations.TopLevelAggregateWithMultipleFields | 6802 | 6802 | 46 | 67 | 1134 |
|  aggregations.NestedAggregation | 16544 | 16544 | 62 | 85 | 2174 |
|  aggregations.AggregationWithWhere | 11970 | 11970 | 51 | 58 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116478 | 22116478 | 2227 | 2304 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116478 | 22116478 | 2303 | 2283 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9701 | 9701 | 33 | 39 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9928 | 9928 | 36 | 36 | 2174 |
|  aggregations.DeeplyNestedCount | 11062184 | 11062184 | 2952 | 2834 | 2008534 |
|  aggregations.InterfacesAggregations | 6242 | 6242 | 49 | 56 | 2080 |
|  aggregations.InterfacesAggregationsWithTwoFields | 11444 | 11444 | 89 | 93 | 2080 |
|  batch-create.BatchCreate | 4200 | 4200 | 129 | 163 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 52 | 57 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 135 | 200 | 3003 |
|  connections.Connection | 14049 | 14049 | 67 | 68 | 2174 |
|  connections.NestedConnection | 41068 | 41068 | 124 | 138 | 4516 |
|  create.SimpleMutation | 7 | 7 | 49 | 66 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 26 | 28 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 232 | 227 | 1040 |
|  delete.NestedDeleteInUpdate | 19165 | 19165 | 198 | 207 | 1183 |
|  2871.NestedRelationshipFilter | 23120 | 23117 | 54 | 59 | 4448 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 36 | 42 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22891 | 22891 | 71 | 86 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 33 | 40 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 49 | 67 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 18 | 24 | 1040 |
|  query.SimpleQueryWithRelationship | 16129 | 16129 | 33 | 42 | 2174 |
|  query.QueryWhere | 9699 | 9677 | 28 | 43 | 2170 |
|  query.SimpleQueryWithNestedWhere | 9866 | 9844 | 43 | 54 | 2170 |
|  query.Nested | 10095502 | 10095502 | 6707 | 6786 | 2008534 |
|  query.NestedWithFilter | 10074351 | 10074351 | 6557 | 6746 | 2004000 |
|  query.OrFilterOnRelationships | 42458 | 42434 | 193 | 173 | 1629 |
|  query.OrFilterOnRelationshipsAndNested | 35723 | 35896 | 186 | 236 | 1629 |
|  query.QueryWithNestedIn | 13254 | 13729 | 50 | 57 | 1180 |
|  query.NestedConnectionWhere | 9801 | 9801 | 46 | 63 | 2174 |
|  query.DeeplyNestedConnectionWhere | 10006 | 9981 | 63 | 79 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 18914 | 18926 | 118 | 144 | 1640 |
|  query.NestedWithRelationshipSingleFilters | 3865 | 3865 | 131 | 193 | 1134 |
|  query.Fulltext | 64 | 64 | 27 | 33 | 16 |
|  query.FulltextWithNestedQuery | 558 | 558 | 39 | 66 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 14059 | 14059 | 38 | 43 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 14059 | 14059 | 55 | 66 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 14204 | 14204 | 55 | 60 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 14204 | 14204 | 93 | 104 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 14929 | 14935 | 91 | 113 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 14404 | 14404 | 94 | 128 | 2174 |
|  sorting.SortMultipleTypes | 3498 | 3498 | 78 | 96 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14504 | 14453 | 96 | 113 | 2174 |
|  sorting.SortOnNestedFields | 14049 | 14049 | 38 | 46 | 2174 |
|  sorting.SortDeeplyNestedFields | 43148 | 43148 | 97 | 93 | 4516 |
|  sorting.ConnectionWithSort | 3281 | 3281 | 72 | 75 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 62 | 65 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 60 | 75 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 317 | 284 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 227 | 256 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 88 | 112 | 2002 |

</details>
        
Old Schema Generation: 25.396s
Schema Generation: 25.355s
Old Subgraph Schema Generation: 29.467s
Subgraph Schema Generation: 28.855s

**Comment by neo4j-team-graphql** [comment]
Thanks for the Neo4j GraphQL Toolbox updates.

The Neo4j GraphQL Toolbox has now been torn down - reopening this PR will republish it.

File: packages/graphql/src/translate/queryAST/ast/fields/OperationField.ts [hunk 1]
```diff
@@ -53,7 +54,13 @@ export class OperationField extends Field {
...
 
-    public isCypherField(): this is this & { operation: CypherOperation | CypherScalarOperation } {
-        return this.operation instanceof CypherOperation || this.operation instanceof CypherScalarOperation;
+    public isCypherField(): this is this & {
+        operation: CypherOperation | CypherScalarOperation | CompositeCypherOperation;
+    } {
+        return (
+            this.operation instanceof CypherOperation ||
+            this.operation instanceof CypherScalarOperation ||
+            this.operation instanceof CypherScalarOperation
```
Comment (a-alle) [inline packages/graphql/src/translate/queryAST/ast/fields/OperationField.ts]: These are the same

File: packages/graphql/src/translate/queryAST/ast/operations/composite/CompositeReadPartial.ts [hunk 2]
```diff
@@ -109,7 +109,8 @@ export class CompositeReadPartial extends ReadOperation {
         if (wherePredicate) {
             matchClause.where(wherePredicate);
         }
-        const subqueries = Cypher.concat(...this.getFieldsSubqueries(nestedContext));
+        const cypherFieldSubqueries = this.getCypherFieldsSubqueries(nestedContext);
```
Comment (a-alle) [inline packages/graphql/src/translate/queryAST/ast/operations/composite/CompositeReadPartial.ts:112]: This variable doesn't really provide much, could be inline in the Cypher.concat

File: packages/graphql/src/translate/queryAST/factory/SortAndPaginationFactory.ts [hunk 3]
```diff
@@ -19,43 +19,59 @@
...
     public createConnectionSortFields(
         options: ConnectionSortArg,
-        entityOrRel: ConcreteEntityAdapter | RelationshipAdapter
+        entityOrRel: ConcreteEntityAdapter | RelationshipAdapter,
+        context: Neo4jGraphQLTranslationContext
     ): { edge: Sort[]; node: Sort[] } {
         if (isConcreteEntity(entityOrRel)) {
-            const nodeSortFields = this.createPropertySort(options.node || {}, entityOrRel);
+            const nodeSortFields = this.createPropertySort({
+                optionArg: options.node || {},
```
Comment (a-alle) [inline packages/graphql/src/translate/queryAST/factory/SortAndPaginationFactory.ts]: should this `|| {}` maybe be in the method itself? I see all callers do the same

File: packages/graphql/src/translate/queryAST/ast/operations/composite/CompositeReadPartial.ts (hunk 2 above)
Comment (MacondoExpress) [reply packages/graphql/src/translate/queryAST/ast/operations/composite/CompositeReadPartial.ts:112]: Just consistency with the other method in the class, that very likely can be unified in a single method, I'll open a separate PR for that!

**Review by a-alle — APPROVED** [review]

**Pull request merged**